### PR TITLE
Assign in i

### LINKED ISSUE #4227: Improve the error message for wrongly-placed `:=`
As an experienced `data.table` user, I can generally recognize the cause of the `:=` error message

> Check that is.data.table(DT) == TRUE. Otherwise, := and `:=`(...) are defined for use in j, once only and in particular ways. See help(\":=\").

However, (a) it still takes a bit of work, and (b) it doesn't fully diagnose the problem, even though it's possible. The following is a proof-of-concept that provides slightly better advice for the common error of providing an update expression to `i = ` instead of `j = `.

Obviously the performance is irrelevant as it is only going to error at this point. 

``` r
library(data.table)
DT <- as.data.table(mtcars)
DT[mpg_per_cyl := mpg/cyl]
#> Error: Check that is.data.table(DT) == TRUE. Otherwise, := and `:=`(...) are defined for use in j, once only and in particular ways. See help(":=").

":=" <- function(x, y) {
  .trace_back <- rlang::trace_back()
  if (length(.trace_back) >= 2L && 
      length(.trace_back[[2]]) >= 3L && 
      as.character(.trace_back[[1L]][[2]][[3]][[1]]) == ":=") {
    stop(":= is intended for use in j only, but is present in i. ", 
         "(Did you forget to put a comma ",
         "before the expression?)")
  } else {
    stop("Check that is.data.table(DT) == TRUE. Otherwise, := and `:=`(...) are defined for use in j, once only and in particular ways. See help(\":=\").")
  }
}

DT[mpg_per_cyl := mpg/cyl]
#> Error: := is intended for use in j only, but is present in i. (Did you forget to put a comma before the expression?)
```

<sup>Created on 2020-02-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #4227 

Nice suggestion @HughParsonage! Indeed it's one of my own more common mistakes/accidents.

### CODE REVIEW

File: R/data.table.R [hunk 1]
```diff
@@ -2611,7 +2611,19 @@ vecseq = function(x,y,clamp) .Call(Cvecseq,x,y,clamp)
...
+  # detect common mistake -- use := in i after forgetting a comma to leave it blank, #4227
+  find_doteq_in_i = function(e) {
+    if (is.call(e)) {
+      if (is.symbol(e[[1L]]) && e[[1L]] == '[.data.table' && is.call(e[[3L]]) && is.symbol(e[[3L]][[1L]]) && e[[3L]][[1L]] == ':=') {
+        stop("Assignment with := is intended for use in j only, but is present in i, the first argument after [. Most often, this happens when forgetting to leave the first argument blank (e.g. DT[newvar := 5] instead of DT[ , new_var := 5]). Please double-check your syntax (see traceback() after this error for another view of this).")
+      }
+      else lapply(e[-1L], find_doteq_in_i)
+    }
+  }
+  for (calli in sys.calls()) find_doteq_in_i(calli)
```
Comment (jangorecki) [inline R/data.table.R]: I have a bad feeling about using `sys.calls` here. I would rather check for `:=` when massaging `isub` argument

**Comment by MichaelChirico** [comment]
I had like that originally, unfortunately if it starts with := the end result looks a bit ugly:

> Error: := ...

so I added the first words as spacing

**Comment by HughParsonage** [comment]
What about `Operator ":=" detected ...`

File: R/data.table.R (hunk 1 above)
Comment (MichaelChirico) [reply R/data.table.R]: I guess this will cover 95% of cases :)

**Comment by MichaelChirico** [comment]
@HughParsonage updated

**Comment by HughParsonage** [comment]
Thanks. I'll just let the approval stand.

**Pull request merged**